CONTRIBUTING.md

```diff
@@ -0,0 +1,179 @@
+USER1_PASSWORD=<ask @plocket>
+SECRET_VAR1=secret-var1-value
+SECRET_VAR2=secret-var2-value
+SECRET_FOR_MISSING_FIELD=secret for missing field
+SECRET_INVALID_THERE_IS_ANOTHER=invalid value for there_is_another
+```
+
+## Running internal tests
+Make the code available to test on the docassemble server:
+```
+npm run setup
+```
+To be able to keep running tests on the same code, avoid takedown.
+
+Use the syntax below to trigger all tests (cucumber and unit tests):
+```
+npm run test
+```
+
+Run only cucumber tests:
+```
+npm run cucumber
+
+# To test Features or Scenarios specific tags:
+npm run cucumber -- "--tags" "@tagname"
+```
+
+To run the unit tests in isolation:
+```
+npm run unit
+```
+
+When done you want to test a new branch or internal test interview files have been updated, clean up before setting up again:
+```
+npm run takedown
+```
+
+# Very general architecture of files and folders
+
+An honest look at our current project architecture - some of our files and folders, what they do, and how they interact.
+
+## Logic architecture
+
+### .feature files
+
+The `.feature` files are written in Gherkin, a syntax cucumber uses. The "code" in there relies on the functions set up in `./lib/steps.js`.
+
+```js
+// The test_something.feature file step
+Then I sign
+
+// relies on the code in steps.js
+Then('I sign', { timeout: -1 }, async () => {
+  return wrapPromiseWithTimeout(
+    scope.steps.sign( scope ),
+    scope.timeout
+  );
+});
+```
+
+### steps.js
+
+`./lib/steps.js` defines all the Gherkin sentences that our users (the developers) can write. See [cucumber documentation on defining steps](https://cucumber.io/docs/cucumber/step-definitions/?lang=javascript). We sometimes use "cucumber expressions" and, but often use regular expressions as we often want to be more flexible about the user's input.
+
+The file handles things like:
+* Creating state
+* Defining steps for the users/developers
+* Defining steps for internal testing, like testing for appropriate errors and error messages
+* Finishing a Scenario
+* Finishing all tests
+
+### scope.js
+
+`./lib/steps.js` makes heavy use of `./lib/scope.js`, where most of the action happens. We have decided to mostly keep it in one file to avoid bouncing back and forth between files.
+
+Many `./lib/steps.js` steps make use of the `scope.steps` object. It is there to avoid duplicating code between steps.
+
+The file hanldes things like:
+* Finding form fields on a page
+* Filling out form fields
+* Generating random data for form fields
+* Waiting for pages to load
+* Checking for server restarts
+* Generating test report content
+
+## Logging
+
+`./lib/utils/log.js` takes care of logging information of various kinds (normal, debug, error) to the console in ways we find useful. We should probably find a logging library instead, but this is what we have for now. An important feature is that it ensures the messages clearly belong to ALKiln and not to some other process.
```

Comment:
```suggestion
`./lib/utils/log.js` takes care of logging information of various kinds (normal, debug, error) to stdout in ways we find useful. We should probably find a logging library instead, but this is what we have for now. An important feature is that it ensures the messages clearly belong to ALKiln and not to some other process.
```

The console kinda implies the browser console, which isn't right.

Reply:
It actually does `console.log()`.

Reply:
Fair, either would be right, since `console.log` in node still goes to `stdout` (and `stderr` apparently, looking into it). IMO it's more important to make it clear that it's a program running on the command line and not javascript running in the browser (which is a reasonable point of confusion).

If you think that the audience (coding knowledgeable folks, might or might not be familiar with node?) would understand "the console" better that fair; I can't speak to folks' perspective internal to node, just the external perspective.